manual/build-toplevel.wiki

```diff
@@ -0,0 +1,26 @@
+= How to build a toplevel =
+
+First, initialize a toplevel using `Js_of_ocaml_compiler.JsooTop.initialize` (if you use js_of_ocaml-toplevel) or `Js_of_ocaml_compiler_dynlink.initialize` (from js_of_ocaml-compiler.dynlink).
+
+Then, build your bytecode program with debug enabled (-g) and linkall (-linkall). You should obviously include all the libraries you want accessible in the final toplevel.
+
+Finaly, compile your toplevel to JavaScript passing the `--toplevel` flags to the js_of_ocaml compiler.
+
+If you want to limit the set of modules available in the toplevel, you can explicitly pass a list of compilation units that should be accessible using the `--export FILE` flag.
+`FILE` must contain names of compilation unit to export - one per line. The `jsoo_listunits` tool, provided by the js_of_ocaml-toplevel package, can be used to generate this list.
+
+For example, the following command will create a file `FILE` containing all compilation unit names provided by the stdlib and the str library.
+{{
+  jsoo_listunits -o FILE stdlib str
+}}
+
+Note that toplevel currently cannot be built using separate compilation
+
+
+= How to build with Dynlink support =
```

Comment:
Re-reading it seems we building something that can dynlink `js_of_ocaml` compiled code. It would be nice to have a full example, how you should compile the main program, how you should compile the plugin. 

Regarding options I still think `--toplevel` is confusing to use when there is no REPL, besides neither `--dynlink` nor `--dynlink-js` are used, what are they used for !?

Finally `js_of_ocaml_compile.dynlink.initialize` could be called by side effect, in the initialisation bits of the library. That way there's no need to change your code. Just link against the library no ?

Reply:
What about 
- `--dynlink-js`, allow to dynlink jsoo-generated js files.
- `--dynlink`, to enable the `Dynlink`  support
-  `--toplevel`, to build a toplevel (similar to `--dynlink` but also embed necessary cmis)

The initialization is now done automatically just by linking the right lib.

Reply:
> --dynlink-js, allow to dynlink jsoo-generated js files.

It's a bit unclear to me what this means exactly. Is this some kind of separate compilation with the `html` file acting as your linker ?  In which order am I supposed to include the files in the html file ? How do the js files interact/see the code in each other ? 

> --dynlink, to enable the Dynlink support

If that's for loading bytecode object files then maybe rather `dynlink-cmo` ?

Reply:
> It's a bit unclear to me what this means exactly. Is this some kind of separate compilation with the html file acting as your linker ? In which order am I supposed to include the files in the html file ? How do the js files interact/see the code in each other ?

It is exactly separate compilation , incrementally loading units.
Separate compilation starts from "just the runtime" and load module in their "linking" other. But you could imagine starting from something bigger. For example, one could start from a program containing all the libs one depends on and only incrementally load `cmo.js`.

Anyway, after some more refactoring:
- I got rid of `--dynlink-js`, this semantic can now be achieved with jsoo `--linkall` flag.
- `--dynlink` now does what's necessary for Dynlink to work

Reply:
Where is separate compilation described ? 

I still think the docs are quite confusing. "How to build with Dynlink support" is rather ambiguous. 

It would be much clearer if you explicitely mentioned what it is possible to dynamically link, i.e. OCaml code compiled to JavaScript or OCaml code compiled to byte code and then in each case have a clear simple example with the steps and flags that are needed.